lib/pqdict.py

```diff
@@ -0,0 +1,185 @@
+from collections import MutableMapping
+
+class _MinEntry(object):
+    """
+    Mutable entries for a Min-PQ dictionary.
+
+    """
+    def __init__(self, dkey, pkey):
+        self.dkey = dkey    #dictionary key
+        self.pkey = pkey    #priority key
+
+    def __lt__(self, other):
+        return self.pkey < other.pkey
+
+class _MaxEntry(object):
+    """
+    Mutable entries for a Max-PQ dictionary.
+
+    """
+    def __init__(self, dkey, pkey):
+        self.dkey = dkey
+        self.pkey = pkey
+
+    def __lt__(self, other):
+        return self.pkey > other.pkey
+
+class PQDict(MutableMapping):
+    def __init__(self, *args, **kwargs):
+        self._heap = []
+        self._position = {}
+        self.update(*args, **kwargs)
+
+    create_entry = _MinEntry     #defaults to a min-pq
+    @classmethod
+    def maxpq(cls, *args, **kwargs):
+        pq = cls()
+        pq.create_entry = _MaxEntry
+        pq.__init__(*args, **kwargs)
+        return pq
+
+    def __len__(self):
+        return len(self._heap)
+
+    def __iter__(self):
+        for entry in self._heap:
+            yield entry.dkey
+
+    def __getitem__(self, dkey):
+        return self._heap[self._position[dkey]].pkey
+
+    def __setitem__(self, dkey, pkey):
+        heap = self._heap
+        position = self._position
+
+        try:
+            pos = position[dkey]
+        except KeyError:
+            # Add a new entry:
+            # put the new entry at the end and let it bubble up
+            pos = len(self._heap)
```

Comment:
you use self._heap and self.heap interchangeably - thats a bad idea. let's just use self._heap everywhere